// @feedloop/qore-react/lib/example.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const qore_client_1 = require("@feedloop/qore-client");
const react_1 = __importDefault(require("react"));
const _1 = __importDefault(require("."));
const qoreClient = new qore_client_1.QoreClient({
    endpoint: "http://localhost:8080",
    organizationId: "FAKE_ORG",
    projectId: "FAKE_PROJECT"
});
qoreClient.init({
    forms: [],
    roles: [],
    tables: [],
    views: [
        {
            id: "allTasks",
            name: "All tasks",
            filter: "",
            parameters: [],
            sorts: [],
            tableId: "tasks",
            fields: [
                {
                    id: "user",
                    name: "user",
                    type: "relation",
                    table: "member",
                    linked: true,
                    multiple: false,
                    deletionProtection: false
                },
                {
                    id: "subtasks",
                    name: "subtasks",
                    type: "relation",
                    deletionProtection: false,
                    table: "subtasks",
                    linked: true,
                    multiple: true
                },
                {
                    id: "done",
                    name: "done",
                    type: "boolean",
                    linked: true,
                    deletionProtection: true
                },
                {
                    type: "text",
                    name: "id",
                    linked: true,
                    id: "id",
                    deletionProtection: false
                },
                {
                    type: "text",
                    name: "name",
                    linked: true,
                    id: "name",
                    deletionProtection: false
                }
            ]
        }
    ]
});
const qoreContext = _1.default(qoreClient);
const View = () => {
    const { data, status } = qoreContext.views.allTasks.useListRow({
        limit: 10,
        offset: 0,
        slug: "slug"
    }, { networkPolicy: "network-and-cache", pollInterval: 1000 });
    qoreContext.views.allTasks.useGetRow("123456");
    qoreContext.views.allTasks.useInsertRow();
    qoreContext.views.allMembers.useInsertRow();
    qoreContext.views.allTasks.useUpdateRow();
    qoreContext.views.allTasks.useDeleteRow();
    return react_1.default.createElement("div", null, status === "loading" && react_1.default.createElement("div", null, "loading..."));
};
const App = () => {
    return react_1.default.createElement(View, null);
};
