// @feedloop/qore-sdk/lib/project/authConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthConfigImpl = void 0;
const url_1 = require("./url");
const common_1 = require("../common");
class AuthConfigImpl {
    constructor(params) {
        this.qore = params.authConfig.qore;
        this.password = params.authConfig.password;
        this._config = params.config;
    }
    async update(values) {
        var _a;
        const id = (_a = this[values.type]) === null || _a === void 0 ? void 0 : _a.id;
        if (!id)
            return;
        await common_1.callApi({
            method: "patch",
            url: url_1.url.authConfig(Object.assign(Object.assign({}, this._config), { projectAuthId: id })),
            data: {
                type: values.type,
                table: values.table,
                identifierField: values.identifierField,
                roleField: values.roleField,
                // @ts-ignore
                passwordField: values.passwordField
            }
        }, this._config.token);
    }
}
exports.AuthConfigImpl = AuthConfigImpl;
