// @feedloop/qore-sdk/lib/project/workflow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowImpl = void 0;
const common_1 = require("../common");
const url_1 = require("./url");
class WorkflowImpl {
    constructor(params) {
        this.id = params.id;
        this.tableId = params.tableId;
        this.name = params.name;
        this.event = params.event;
        this.type = params.type;
        this.active = params.active;
        this.actions = params.actions;
        this.preCondition = params.preCondition;
        this._config = params.config;
    }
    delete() {
        return common_1.callApi({
            method: "delete",
            url: url_1.url.workflow(Object.assign(Object.assign({}, this._config), { workflowId: this.id }))
        }, this._config.token);
    }
    update(workflow) {
        return common_1.callApi({
            method: "patch",
            url: url_1.url.workflow(Object.assign(Object.assign({}, this._config), { workflowId: this.id })),
            data: workflow
        }, this._config.token);
    }
}
exports.WorkflowImpl = WorkflowImpl;
