// @feedloop/qore-sdk/lib/user/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectImpl = void 0;
const common_1 = require("../common");
class ProjectImpl {
    constructor(params) {
        this.id = params.id;
        this.name = params.name;
        this._orgId = params.orgId;
        this._url = params.url;
        this._token = params.userToken;
    }
    async get() {
        await common_1.callApi({
            method: "get",
            url: this._url.project(this._orgId)
        }, this._token);
    }
    async tables(props = {}) {
        const { nodes } = await common_1.callApi({
            method: "get",
            url: this._url.project(this._orgId, this.id) + "/tables"
        }, this._token);
        return nodes;
    }
    async delete() {
        await common_1.callApi({
            method: "delete",
            url: this._url.project(this._orgId, this.id)
        }, this._token);
    }
    async update(project) {
        await common_1.callApi({
            method: "patch",
            url: this._url.project(this._orgId, this.id),
            data: project
        }, this._token);
    }
}
exports.ProjectImpl = ProjectImpl;
